### ISSUE TITLE
Server will Panic on Server:initialise_packet_from_bytes when parsing Auth packet generated by radclient

### ISSUE THREAD

**Issue by CoderChristopher** [MikhailMS/rust-radius#28]
Hello,

I was running the example code for an async server, testing authorization with various parameters using the radclient tool when I came across this combination that raised a panic:

radclient cmd:
`echo 'Message-Authenticator = 0x00, User-Name = user, CHAP-Password = password' | radclient -x 127.0.0.1 auth secret`

The sent text as reported by radclient -x:
```
Sent Access-Request Id 17 from 0.0.0.0:cf18 to 127.0.0.1:1812 length 63
        Message-Authenticator = 0x00
        User-Name = "user"
        CHAP-Password = 0x09d50e1b2b3a578d394372ccf785117458
        Cleartext-Password = "password"
```

Panic Result in handle_auth_request:
```
thread 'tokio-runtime-worker' panicked at /home/chris/.cargo/registry/src/index.crates.io-6f17d22bba15001f/radius-rust-0.4.2/src/protocol/radius_packet.rs:394:37:
slice index starts at 65 but ends at 63
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: task 9 panicked
```

I examined the code that was raising the panic:
```
    pub fn initialise_packet_from_bytes(dictionary: &Dictionary, bytes: &[u8]) -> Result<RadiusPacket, RadiusError> {
        let code           = TypeCode::from_u8(bytes[0])?;
        let id             = bytes[1];
        let authenticator  = bytes[4..20].to_vec();
        let mut attributes = Vec::new();

        let mut last_index = 20;

        while last_index != bytes.len() {
            let attr_id     = bytes[last_index];
            let attr_length = bytes[last_index + 1] as usize;
            //Right here is the issue\/
            let attr_value  = &bytes[(last_index + 2)..=(last_index + attr_length - 1)];
            //This is the issue ^

            match RadiusAttribute::create_by_id(dictionary, attr_id, attr_value.to_vec()) {
                Some(attr) => {
                    attributes.push(attr);
                    last_index += attr_length;
                },
                _          => return Err( RadiusError::MalformedPacketError {error:format!("attribute with ID: {} is not found in dictionary", attr_id)} )
            }
```
Then I added some quick and dirty debugging to the above code:
`            println!( "DEBUG {last_index} {attr_id} {attr_length}" );`

And got this output:
```
Async Radius Server starting...
DEBUG: Authenticating on 127.0.0.1:1812
DEBUG: Accounting on 127.0.0.1:1813
DEBUG: COA on 127.0.0.1:3799
DEBUG Handling CoA Request
DEBUG Handling ACCT Request
DEBUG Handling AUTH Request
DEBUG 20 80 18
DEBUG 38 1 6
DEBUG 44 3 19
DEBUG 63 0 0
thread 'tokio-runtime-worker' panicked at /home/chris/.cargo/registry/src/index.crates.io-6f17d22bba15001f/radius-rust-0.4.2/src/protocol/radius_packet.rs:394:37:
slice index starts at 65 but ends at 63
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: task 9 panicked
```

I am not very well versed in the Radius protocol, but from what I gather the client is sending an attribute with length 0 which caused the array lookup to be invalid.

I did a naive patch of the following:

```
        while last_index != bytes.len() {
            let attr_id     = bytes[last_index];
            let attr_length = bytes[last_index + 1] as usize;

            println!( "DEBUG {last_index} {attr_id} {attr_length}" );
            //\/my patch
            if attr_length == 0 {
                break;
            }
            //^my patch

            let attr_value  = &bytes[(last_index + 2)..=(last_index + attr_length - 1)];

```

Which seemed to allow the request to process successfully:
```
Async Radius Server starting...
DEBUG: Authenticating on 127.0.0.1:1812
DEBUG: Accounting on 127.0.0.1:1813
DEBUG: COA on 127.0.0.1:3799
DEBUG Handling CoA Request
DEBUG Handling ACCT Request
DEBUG Handling AUTH Request
DEBUG 20 80 18
DEBUG 38 1 6
DEBUG 44 3 19
DEBUG 63 0 0
Received Request from 127.0.0.1:37251
Request RadiusPacket { id: 20, code: AccessRequest, authenticator: [13, 11, 127, 107, 247, 218, 204, 175, 223, 92, 42, 224, 239, 22, 137, 154], attributes: [RadiusAttribute { id: 80, name: "Message-Authenticator", value: [172, 12, 181, 79, 162, 127, 226, 210, 68, 80, 50, 179, 21, 149, 105, 46] }, RadiusAttribute { id: 1, name: "User-Name", value: [117, 115, 101, 114] }, RadiusAttribute { id: 3, name: "CHAP-Password", value: [178, 91, 23, 173, 45, 231, 93, 112, 197, 148, 21, 209, 199, 209, 19, 62, 72] }] }
DEBUG Handling AUTH Request
```

But I have no idea what the correct solution should be or what is appropriate.

I am happy to offer more input as needed.

Thanks!
Chris C.

**Comment by MikhailMS** [comment]
Hi there @CoderChristopher  

Thank you for rasing this issue - I'll give it better look tomorrow, but atm I am thinking whether it is aligned with RADIUS RFCs - afaik Message-Authenticator should be 18 octets long

So I'd try to find exact info and then see what should be the way to handle cases when it's empty/of not expected length

**Comment by MikhailMS** [comment]
@CoderChristopher hello there,

I am currently looking through the RFCs and while doing so - would you be able to get the raw bytes of the message that panics? (smth like [here](https://github.com/MikhailMS/rust-radius/blob/master/src/protocol/radius_packet.rs#L610), or like line 630 to be more precise)

This way I'd be able add them to the test suite and check against as I am figuring out the way forward :)

**UPD**:
If I am reading [this bit of RFC2865](https://datatracker.ietf.org/doc/html/rfc2865#section-5) correctly, no attribute can have a length of zero, because length of the attribute always includes:
* Type of the attribute, which is 1 octet AND
* Length of the attribute, which is 1 octet as well
* Value of the attribute could be zero (or more) octets

So the length of the attribute should be at least 2 octets and not 0

Reading the same part of the RFC, it suggestes (as in SHOULD) to return Access-Reject; but there is nothing mandatory for this case in the RFC

So I am proposing to catch this error state and return RadiusError (func already does it, but not for this case) out to caller to handle: silently ignore or return Access-Reject instead of panicing :)

Would that work for you or you are aware of the legit situation where this is the case and software behaves differently to this lib?

**Comment by CoderChristopher** [comment]
Here are the raw bytes sent from radclient right off the network. After reading [RFC2865 section 3](https://datatracker.ietf.org/doc/html/rfc2865#section-3) and referencing https://www.iana.org/assignments/radius-types/radius-types.xhtml I manually decoded the packet:

```
[
    //1=Access-Request, 102=identifier, 063=length of packet
    1, 102, 0, 63, 

    //Authenticator
    127, 191, 122, 67, 238, 159, 126, 219, 178, 175, 203, 71, 168, 246, 183, 156,

    //80=Message-Authenticator, 18=length (https://www.rfc-editor.org/rfc/rfc2869.html#section-5.14)
    80, 18, 212, 157, 60, 72, 7, 79, 228, 174, 119, 139, 180, 99, 124, 31, 44, 141,

    //1=User-Name, 6=length "user"
    1, 6, 117, 115, 101, 114,

    //3=CHAP-Password, 19=length (https://www.rfc-editor.org/rfc/rfc2865.html#section-5.3)
    3, 19, 36, 157, 0, 194, 19, 36, 235, 42, 20, 152, 135, 215, 72, 137, 191, 151, 195
]

```
However, after looking at this, it leaving me scratching my head a bit.

From my manual breakdown we can see there are three distinct attributes, but based upon my DEBUG result above we see there are 4 iterations of the processing loop in initialise_packet_from_bytes().

Looking at my handling code for the auth_requests:
```
eprintln!( "DEBUG Handling AUTH Request" );
    let mut request = [0u8; 4096];

    let ( _, source_addr) = self.auth_socket.recv_from(&mut request).await?;

    let request_packet = match Server::initialise_packet_from_bytes( &self.server, &request ) {
        Ok( packet ) => packet,
        Err( error ) => {
            eprintln!( "Error: {error}" );
            continue;
        }
    };
```
We see that my request buffer is 0 initialized with 4096 bytes, then the body of the udp packet is dumped into it, then it gets parsed by the initalise_packet_from_bytes() function.

Looking at my debug out against your function:

_DEBUG Handling AUTH Request
DEBUG 20 80 18
DEBUG 38 1 6
DEBUG 44 3 19
DEBUG 63 0 0_

```
pub fn initialise_packet_from_bytes(dictionary: &Dictionary, bytes: &[u8]) -> Result<RadiusPacket, RadiusError> {
        let code           = TypeCode::from_u8(bytes[0])?;
        let id             = bytes[1];
        let authenticator  = bytes[4..20].to_vec();
        let mut attributes = Vec::new();

        let mut last_index = 20;

        while last_index != bytes.len() {
            let attr_id     = bytes[last_index];
            let attr_length = bytes[last_index + 1] as usize;

            println!( "DEBUG {last_index} {attr_id} {attr_length}" );
            //\/my patch
            if attr_length == 0 {
                break;
            }
            //^my path

            let attr_value  = &bytes[(last_index + 2)..=(last_index + attr_length - 1)];

            match RadiusAttribute::create_by_id(dictionary, attr_id, attr_value.to_vec()) {
                Some(attr) => {
                    attributes.push(attr);
                    last_index += attr_length;
                },
                _          => return Err( RadiusError::MalformedPacketError {error:format!("attribute with ID: {} is not found in dictionary", attr_id)} )
            }
        }

        let mut packet = RadiusPacket{
            id,
            code,
            authenticator,
            attributes:    Vec::new()
        };
        packet.set_attributes(attributes);

        Ok(packet)
    }
```
We can see the first three DEBUG prints correspond to the first three attributes of the sent packet. But the fourth time around the while loop is starting to read past the end of the formal data section of the Radius packet and into the 0 initialized bytes. 

Throwing in one more hacked in DEBUG println:

```
pub fn initialise_packet_from_bytes(dictionary: &Dictionary, bytes: &[u8]) -> Result<RadiusPacket, RadiusError> {
    let code           = TypeCode::from_u8(bytes[0])?;
    let id             = bytes[1];
    let authenticator  = bytes[4..20].to_vec();
    let mut attributes = Vec::new();

    let mut last_index = 20;

    //\/ My Added
    let bytes_len = bytes.len();
    println!( "Size of bytes: {bytes_len}" );
    //^ Added debug
    while last_index != bytes.len() {
```

We get the debug output:
```
Size of bytes: 4096                                                                                                                                            
DEBUG 20 80 18                                                                                                                                                 
DEBUG 38 1 6                                                                                                                                                   
DEBUG 44 3 19                                                                                                                                                  
DEBUG 63 0 0          
```                                                                                                                                         

As we can see, because the while is taking its size from the size of the recipient buffer its going to continue to run over.

However, I was reading RFC 2865 tonight and [noting section 3](https://datatracker.ietf.org/doc/html/rfc2865#section-3):

> 
> A summary of the RADIUS data format is shown below.  The fields are
>    transmitted from left to right.
> [...]
>    Length
> 
>       The Length field is two octets.  It indicates the length of the
>       packet including the Code, Identifier, Length, Authenticator and
>       Attribute fields.  **Octets outside the range of the Length field
>       MUST be treated as padding and ignored on reception**.  If the
>       packet is shorter than the Length field indicates, it MUST be
>       silently discarded.  The minimum length is 20 and maximum length
>       is 4096.

This suggests to me that the size of reading should be taken directly from bytes [2-3] of the radius packet (in this case values 00 & 63 ) and used as the reading length for the while loop. If there is any data past that the specification explicitly says to disregard that data. 

Putting my money where my mouth is and patching the function like so:
```
    /// Initialises RADIUS packet from raw bytes
    pub fn initialise_packet_from_bytes(dictionary: &Dictionary, bytes: &[u8]) -> Result<RadiusPacket, RadiusError> {
        let code           = TypeCode::from_u8(bytes[0])?;
        let id             = bytes[1];
        //\/my new patch
        let length         = bytes[2] as usize + bytes[3] as usize;
       //^my new patch
        let authenticator  = bytes[4..20].to_vec();
        let mut attributes = Vec::new();
        

        let mut last_index = 20;
        //\/patched
        while last_index != length {
        //^Patched
            let attr_id     = bytes[last_index];
            let attr_length = bytes[last_index + 1] as usize;     
```   
Yields a now happily running server handling auth requests :)

I would humbly submit that this is the correct solution according to the spec.

**Comment by CoderChristopher** [comment]
I note a small bug in this the line:
`let length         = bytes[2] as usize + bytes[3] as usize;`

Will not render the correct results if byte 2 is anything other than zero.

Code that does work, but that is a bit ugly:
`let length         = (bytes[2] as usize + 255) + bytes[3] as usize;`

**Comment by CoderChristopher** [comment]
Okay, maybe I got a little ahead of myself.

Just found a bug with my proposed 'solution'.

If the byte[2] is anything other than 0 the length will be wrong. It needs to be multiplied by 256 to be correct:

```
pub fn initialise_packet_from_bytes(dictionary: &Dictionary, bytes: &[u8]) -> Result<RadiusPacket, RadiusError> {
    let code           = TypeCode::from_u8(bytes[0])?;
    let id             = bytes[1];
    let length         = (bytes[2] as usize * 256) + bytes[3] as usize;
    let authenticator  = bytes[4..20].to_vec();

```

**Comment by MikhailMS** [comment]
@CoderChristopher appreciate your input on the matter - that actually a great spot, I missed it when was reading in yesterday :) 
Surprised that it was not getting the packet size from the packet itself and was relying on the `bytes.len()`

There is a [function](https://github.com/MikhailMS/rust-radius/blob/development/src/tools/mod.rs#L400) within `tools` module to convert from bytes to u16 (and then we can safely cast it to usize), so I'd fallback to that function instead, but again, appreciate your code suggestion

Later this week I'd be making a new release based on this conversation which:
1. Makes use of the length mentioned within packet (as per your suggestion and, as you noted, per [RFC2865](https://datatracker.ietf.org/doc/html/rfc2865#section-5))
2. Returns an error if attribute length is actually zero (as per [RFC2865](https://datatracker.ietf.org/doc/html/rfc2865#section-5))

As a side note, this could also be avoided by doing [this](https://github.com/MikhailMS/rust-radius/blob/development/examples/sync_radius_server.rs#L97) - but obviously having it properly handled by library is better and such would be done

**Comment by CoderChristopher** [comment]
Awesome, I am glad we could dig in and find the best solution that helps the project align closer to the RFC.

And happy to be of service, my first interaction with an open source project, so thank you for being easy to work with!

I might take a look at some of the issues flagged for fixing. Never committed to an open source project and cannot guarantee I will have time, but happy to help if I am able.

**Comment by MikhailMS** [comment]
@CoderChristopher, thank you for kind words

I do appreicate help on the project, so if you feel like contributing - you are more than welcome to tackle any open issue, try and break existing code (in case something else like this issue exists) and flag it or there maybe some docs missing :)

**Comment by MikhailMS** [comment]
Fixed by [ac13470](https://github.com/MikhailMS/rust-radius/commit/ac13470c74e0de45f52dc90d797a8383a19b0ac8)
